### ISSUE TITLE
Apparent deadlock when a channel is configured to write to itself

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#2415]
Setup a channel with a JS reader that reads messages into the channel. Set the destination as a channel writer that writes to itself. Deploy the channel, then attempt to stop the channel. Stopping should hang.

There may be a deadlock happening in Donkey somewhere when you have a channel that writes to itself. If possible, it would be great to detect and prevent this deadlock before it happens.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-2479
Reporter: brentm
Created: 2013-04-04T08:15:48.000-0700

**Comment by rbeckman-nextgen** [comment]
People DO actually use channels that route to themselves. Sometimes it's in conjunction with a filter (obviously to prevent an infinite loop), or sometimes it's splitting a batch up into multiple messages in a preprocessor (where there's not really any chance of an infinite loop because once a batch is split up, the discrete messages aren't going to be split up again). It's more likely than you think...

Anyway, currently this is because channels use a single thread executor, so when the VmDispatcher calls dispatchRawMessage on the same channel, a new MessageTask gets submitted. But that task is waiting on the original task to complete, and the original task is waiting on the new task to complete (EVEN if the message should be queued on the source, a DispatchResult still has to be returned).

The quick fix here is to disallow channels routing to themselves, but we'll have to discuss what we want to have happen.

Imported Comment. Original Details:
Author: narupley
Created: 2013-04-04T08:35:04.000-0700

**Comment by rbeckman-nextgen** [comment]
Possible solution: Only allow a channel to route to itself if its source queue is on. All tasks submitted would just drop in the queue, and one task wouldn't be dependent on another.

This is an easy change for the VmDispatcher because it has its channel ID; it just needs to get the deployed channel from the engine controller and see whether respondAfterProcessing is false on the source connector. Or, we could circumvent that altogether and just not allow the channel itself to appear in the Channel Writer combo box when the source queue is turned off.

On the VMRouter side, we would have to add a channel ID to its constructor (signifying the channel it's being used in), and then when it actually does the routing, we would check A) whether the channel it's routing to is the same one that it's being used in, and B) if the source queue is off. If both of those are true, it would throw an exception with some message like "possible deadlock detected, source queue is off".

I think that would solve this problem. The only danger area is the global deploy / shutdown script... maybe we just make the router unavailable in those scripts.

Imported Comment. Original Details:
Author: narupley
Created: 2013-04-04T09:56:47.000-0700

**Comment by rbeckman-nextgen** [comment]
Revision 6417: Fixed deadlock problem when a channel is configured to write to itself (either directly or through some sort of multi-channel cycle). The Channel Writer now has a response timeout option, defaulted to 10 seconds. When a message is sent through VmDispatcher, if the timeout is greater than zero, a DispatchTask is submitted and blocked on for the specified amount of time, after which a TimeoutException is thrown. Otherwise (if the timeout is zero), the call to dispatchRawMessage is just made directly as before. Similar logic has been added to VMRouter, which also now has new overloaded composites for routeMessage and routeMessageByChannelId which allow for String vs. RawMessage and timeout vs. no timeout.

Imported Comment. Original Details:
Author: narupley
Created: 2013-05-29T17:53:37.000-0700

**Comment by rbeckman-nextgen** [comment]
Halting a channel that is configured to write to itself will allow it to break the deadlock.  If a response timeout is hit, the error message explains that a cycle may be present and the channel continues processing after the message errors because a response was not received.

Imported Comment. Original Details:
Author: jacobb
Created: 2013-06-27T15:56:15.000-0700

**Comment by rbeckman-nextgen** [comment]
This can cause threading issues if the user sets a low timeout and sends a lot of messages.

Imported Comment. Original Details:
Author: narupley
Created: 2013-07-18T11:30:37.000-0700

**Comment by rbeckman-nextgen** [comment]
Revision 6690: Reverted the addition of a response timeout to the Channel Writer.

This is too dangerous to add right now because if a user specified a response timeout of 1 ms and then blasted the channel with messages, the number of threads queued up at the target channel waiting to acquire a lock would grow unbounded.

A better solution would be to repurpose VMRouter so that it acts as an actual Donkey DestinationConnector, with queuing and everything (MIRTH-2780).

Imported Comment. Original Details:
Author: narupley
Created: 2013-07-19T10:50:16.000-0700

**Comment by rbeckman-nextgen** [comment]
The deadlock still occurs now but it is possible to halt the channel and break the deadlock. For now this is considered user error. If a user wants a channel to write to itself, the source or destination queue should be enabled.

Imported Comment. Original Details:
Author: wayneh
Created: 2013-08-06T17:10:47.000-0700
